### PR TITLE
Prysm_libP2P_Implementation_project_proposal

### PR DESCRIPTION
**Pull request by rose2221**
Could you please review it @taxmeifyoucan

### CODE REVIEW

**Comment by rose2221** [comment]
Thanks @taxmeifyoucan ,This proposal was reviewed and contributed by everyone. After our discussions, I uploaded the final version.

File: projects/Prysm libP2P Implementation [hunk 1]
```diff
@@ -0,0 +1,86 @@
+# Prysm libP2P Implementation | Design Doc
```
Comment (shyam-patel-kira) [inline projects/Prysm libP2P Implementation:1]:
```suggestion
# Custom golang implementation of libp2p in Prysm
```

File: projects/Prysm libP2P Implementation [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+Custom go-libP2P implementation with specific components as used by Prysm.
```
Comment (shyam-patel-kira) [inline projects/Prysm libP2P Implementation:2]:
```suggestion
An in-house implementation of the necessary parts of the libp2p protocol in the Prysm client
```

File: projects/Prysm libP2P Implementation [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+## Motivation
+The project aims to develop an in-house implementation of p2p communication library and leverage some of the core components from the existing [go-libp2p](https://github.com/libp2p/go-libp2p) libraries. go-libP2P library(and its variants in other languages) are used by almost all consensus clients for p2p network communications among beacon chain nodes.The scope of the project enables the Prysm team to be independent of any third party for an integral components. It also allows for elimination of redundant components from go-libp2p which are not actively used, while achieving the same performance. Implementing the project also involves incorporating a deep understanding of libP2P and networking layer.
```
Comment (shyam-patel-kira) [inline projects/Prysm libP2P Implementation:5]:
```suggestion
The project aims to develop an in-house implementation of p2p communication library and leverage some of the core components from the existing [go-libp2p](https://github.com/libp2p/go-libp2p) libraries. go-libP2P library(and its variants in other languages) are used by almost all consensus clients for p2p network communications among beacon chain nodes. The scope of the project enables the Prysm team to be independent of any third party for the integral components. It also allows for elimination of redundant components from go-libp2p which are not actively used, while achieving the same performance. Implementing the project also involves incorporating a deep understanding of libP2P and networking layer. 
```

File: projects/Prysm libP2P Implementation [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+## Project Description
+The libP2P protocol has several components for example: Noise, multiplexer, pubsub, ping etc. Not all these components are used by Prysm, hence the major tasks or the solution for the problem at hand are:
+* Segregate the components in libP2P which are used in Prysm, this involves a clear marking of:
+    *  the individual elements which can be used as-it-is 
+    *  the elements which need refactoring to incorporate only the segments used by libP2P
+    *   the components which can be eliminated completely.
```
Comment (shyam-patel-kira) [inline projects/Prysm libP2P Implementation:12]:
```suggestion
    *  the components which can be eliminated completely. 
```

File: projects/Prysm libP2P Implementation [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+## Project Description
+The libP2P protocol has several components for example: Noise, multiplexer, pubsub, ping etc. Not all these components are used by Prysm, hence the major tasks or the solution for the problem at hand are:
+* Segregate the components in libP2P which are used in Prysm, this involves a clear marking of:
+    *  the individual elements which can be used as-it-is 
+    *  the elements which need refactoring to incorporate only the segments used by libP2P
+    *   the components which can be eliminated completely. 
+    *   Thanks to @MaxDav for noting some required changes in his [notes](https://hackmd.io/zIWLqRzWT76I5T_sPbJ0KA).
```
Comment (shyam-patel-kira) [inline projects/Prysm libP2P Implementation:13]:
```suggestion
    *  Thanks to @MaxDav for noting some required changes in his [notes](https://hackmd.io/zIWLqRzWT76I5T_sPbJ0KA).
```

File: projects/Prysm libP2P Implementation [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+* [`mplex`](https://github.com/libp2p/specs/tree/master/mplex)
+* [`yamux`](https://github.com/libp2p/specs/blob/master/yamux/README.md)
+* [`gossipsub`](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.0.md)
+* [`ssz`](https://github.com/ethereum/consensus-specs/blob/dev/ssz/simple-serialize.md) encoding 
+
+
+## Roadmap
+
+**Month 1: Design and Initial Development** 
+* Deep dive into the p2p-interface.md from the consensus specs. Understand the theoretical aspects thoroughly.
```
Comment (shyam-patel-kira) [inline projects/Prysm libP2P Implementation:37]:
```suggestion
- Deep dive into the p2p-interface.md from the consensus specs. Understand the theoretical aspects thoroughly.
```

File: projects/Prysm libP2P Implementation [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+* Start examining the existing libp2p components used in the Ethereum beacon chain. Identify which components are utilized by Prysm and which are not necessary.
```
Comment (shyam-patel-kira) [inline projects/Prysm libP2P Implementation:38]:
```suggestion
- Start examining the existing libp2p components used in the Ethereum beacon chain. Identify which components are utilized by Prysm and which are not necessary.
```

**Review by shyam-patel-kira — CHANGES_REQUESTED** [review]

**Comment by shyam-patel-kira** [comment]
Hey @rose2221, thanks for the PR, just requested some nits

**Review by shyam-patel-kira — APPROVED** [review]

**Comment by shyam-patel-kira** [comment]
@taxmeifyoucan this can be merged now

**Pull request merged**